visualization/pg_joins.js

```diff
@@ -0,0 +1,197 @@
+    const res= (await client.query(query));
+    return res.rows[0]["id"]
+  }
+  catch(e){
+    console.log(e);
+  }
+}
+
+const fetch_sitesId= async (client, sitename)=>{
+  try{
+    console.log(sitename)
+    const query=format('SELECT id FROM sites where sitename = %L',sitename)
+    const res= (await client.query(query));
+    return res.rows[0]["id"]
+  }
+  catch(e){
+    console.log(e);
+  }
+}
+
+const fetch_specieId= async (client, specie)=>{
+  try{
+    const res= (await client.query('SELECT id FROM species WHERE scientificname = $1', [specie]));
+    return res.rows[0]["id"]
+  }
+  catch(e){
+    console.log(e);
+  }
+}
+
+
+const getTable1=async (cultivars, sites_cultivars, sites) => {
+  const client = await pool.connect()
+  try{
+    await client.query('BEGIN')
+    let dat= cultivars.map(async data =>{ 
+        let specie=data.pop();
+        let specie_id = await fetch_specieId(client,specie)
+        data.push(parseInt(specie_id))
+        return data
+      })
+    let t1= ((await client.query("SELECT UTC_NOW()")).rows[0]["utc_now"]);
+    let res=  Promise.all(dat).then(async (data)=>{ 
+                const query1 = format('INSERT INTO cultivars(name,ecotype,notes,specie_id) VALUES %L',data)
+                await client.query(query1)
+              }).catch((e)=> console.log(e))
+              .then(async()=>{
+                  sites.map((site)=>{
+                  site.pop()
+                  site.pop()
+                  site.pop()
+                  site.push("some text")
+                })
+                const query3 = format('INSERT INTO sites(sitename,city,state,country,notes,time_zone,soil,soilnotes) VALUES %L',sites)
+                await client.query(query3)
+                  let sites_data=sites_cultivars.map(async data =>{
+                  let cultivar_name= data.shift();
+                  let specie_id= data.shift();
+                  let sitename= data.shift();
+                  let cultivar_id= await fetch_cultivarsId(client, cultivar_name, specie_id);
+                  let site_id= await fetch_sitesId(client,sitename);
+                return [site_id, cultivar_id]
+                  })
+                  return Promise.all(sites_data)
+                })
+                .then(async (sites_data)=>{
+                      const query2 = format('INSERT INTO sites_cultivars(site_id,cultivar_id) VALUES %L',sites_data)
+                      await client.query(query2)
+                  })
+                  .then(async()=>{
+                    let join=format("SELECT name as cultivar, sitename as plot FROM cultivars JOIN sites_cultivars ON cultivars.id = sites_cultivars.cultivar_id AND cultivars.created_at >= timestamptz %L JOIN sites ON sites_cultivars.site_id = sites.id",t1)
+                    res=await client.query(join);                    
+                    await client.query('ROLLBACK')
```

Comment:
Why the ROLLBACK here and elsewhere in this file? I'm not sure I understand the purpose of this file if it's kept.

Reply:
As here we just need the data after performing joins on some tables, for only visualization purposes and every other operation performed on the database before taking table joins (like `INSERT`)  must not be applied to the database. For insertion purposes, we have YABA API which can perform data ingestion after a user has validated the data. Therefore, a 
`ROLLBACK` operation is performed at last after taking joins.
You can also take a look at this issue: #41

Reply:
Thank you for explaining. Can you please comment the code to that effect so that it's clearer in the future?